### PR TITLE
Fix imports not being removed from MAS files.

### LINKED ISSUE #30327: MetadataAsSource isn't aware of nullability
The output from MetadataAsSource is lacking nullability annotations (`?`).
The `[Nullable]` attribute is displayed, instead of the `?` in the proper location(s).

image

Note: we need to display `#nullable enable/disable` which can be tricky

FYI @dpoeschl

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/44200
Fixes #30327

Also does a reasonable effort of emitting `#nullable enable/disable` directives in MAS files when it contains signatures wiht types that are annotated/oblivious.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

File: src/Features/CSharp/Portable/MetadataAsSource/FormattingRule.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Utilities;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Formatting.Rules;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.MetadataAsSource
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MetadataAsSource/FormattingRule.cs:12]: this is justa move.

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 2]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
                 new CSharpParenthesizedPatternReducer(),
                 new CSharpDefaultExpressionReducer());
 
-        private class FormattingRule : AbstractMetadataFormattingRule
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:74]: this type was moved to a sibling file.

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 3]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
...
+        ///                 #nullable disable
+        ///                 string bar
+        ///                 #nullable enable
+        ///                 string baz);
+        /// </code>
+        /// 
+        /// Then we would be unable to handle that.  However, this is highly unlikely to happen, and so we accept the
+        /// inaccuracy for the purpose of simplicity and for handling the much more common cases of either the entire
+        /// file being annotated, or the user individually disabling annotations at the member level.
+        /// </summary>
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:97]: i discussed the above algorithm with @jcouv .  We think it should meet the needs of like 99%+ of all code.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Does this also close #30327?

**Comment by JoeRobich** [comment]
@CyrusNajmabadi So the extra using directives were being added because we were always adding Attribute namespaces to the list of usings, and the .NET 5 assemblies now have nullable annotations?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Does this also close #30327?

Yup!

**Comment by CyrusNajmabadi** [comment]
>@CyrusNajmabadi So the extra using directives were being added because we were always adding Attribute namespaces to the list of usings, and the .NET 5 assemblies now have nullable annotations?

That was one problem.  But the primary problem was that we basically turned off the step to then clean up those imports.  So any added imports that turned out to be unnecessary just were not removed.  Nullable just exacerbated this as you might have a ton of those in code which just exploded this sort of thing.

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs [hunk 4]
```diff
@@ -2352,9 +2336,284 @@ void M([|D|]&lt;int&gt; lambda)
...
+// {CodeAnalysisResources.InMemoryAssembly}
+#endregion
+
+#nullable enable
+
+public class TestType
+{{
+    public TestType();
+
+#nullable disable
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2448]: When looking at the implementation at first this seemed a bit strange to tag it this way but I have to say I think I actually like it -- it means there's always a #nullable enable at the top which is easy to see, and the special cases are called out specifically.

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs [hunk 5]
```diff
@@ -2352,9 +2336,284 @@ void M([|D|]&lt;int&gt; lambda)
...
+#nullable enable
+
+using System;
+
+public class TestType
+{{
+    public TestType();
+
+    public void [|M1|](ICloneable s);
+}}";
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: Do we not generate nested types?

Comment (jasonmalinowski) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: Anything needed for testing with a namespace?

Comment (jasonmalinowski) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: You had the code for "once we're done with a type, restore it back to "enabled" which I presume means yes, we do have that case. If not, then perhaps that code is unnecessary?

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs [hunk 6]
```diff
@@ -591,18 +591,14 @@ internal class AllowNullAttribute : System.Attribute { }
 // {CodeAnalysisResources.InMemoryAssembly}
 #endregion
 
-using System.Runtime.CompilerServices;
-
-[NullableContextAttribute(1)]
-public interface [|C|]<[NullableAttribute(2)] T>
+public interface [|C|]<T>
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:594]: not that i can tell.  @jcouv we dont' need `#nullable enable` here right?  we don't have an annotated reference type afaict.

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2448]: yup.  that seemed really sane and sensible.  it's the coding pattern i would prefer to see tbh :)

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: i can add test.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: i'll add test.

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:97]: great!

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 7]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
                 new CSharpParenthesizedPatternReducer(),
                 new CSharpDefaultExpressionReducer());
 
-        private class FormattingRule : AbstractMetadataFormattingRule
+        /// <summary>
+        /// Adds <c>#nullable enable</c> and <c>#nullable disable</c> annotations to the file as necessary.  Note that
+        /// this does not try to be 100% accurate, but rather it handles the most common cases out there.  Specifically,
+        /// if a file contains any nullable annotated/not-annotated types, then we prefix the file with <c>#nullable
+        /// enable</c>.  Then if we hit any members that explicitly have *oblivious* types, not no annotated or
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:80]:
```suggestion
        /// enable</c>.  Then if we hit any members that explicitly have *oblivious* types, but no annotated or
```

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 8]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
...
+        /// file being annotated, or the user individually disabling annotations at the member level.
+        /// </summary>
+        protected override async Task<Document> AddNullableRegionsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var options = (CSharpParseOptions)tree.Options;
+
+            // Only valid for C# 8 and above.
+            if (options.LanguageVersion < LanguageVersion.CSharp8)
+                return document;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:105]: The problem with that is that you then get an error:

image

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 9]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
...
-            protected override AdjustNewLinesOperation GetAdjustNewLinesOperationBetweenMembersAndUsings(SyntaxToken token1, SyntaxToken token2)
+            return (HasAnnotation(node, NullableSyntaxAnnotation.None),
+                    HasAnnotation(node, NullableSyntaxAnnotation.Annotated),
+                    HasAnnotation(node, NullableSyntaxAnnotation.NotAnnotated));
+        }
+
+        private bool HasAnnotation(SyntaxNode node, SyntaxAnnotation annotation)
+        {
+            // see if any child nodes have this annotation.  Ignore anything in attributes (like `[Obsolete]void Goo()`
+            // as these are not impacted by `#nullable` regions.  Instead, we only care about signature types.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:131]: yeah, was a struggle to figure out how to doc  best

File: src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs [hunk 10]
```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
...
+
+            var currentlyEnabled = true;
+
+            foreach (var member in type.Members)
+            {
+                if (member is BaseTypeDeclarationSyntax)
                 {
-                    return null;
+                    // if we hit a type, and we're currently disabled, then switch us back to enabled for that type.
+                    // This ensures whenever we walk into a type-decl, we're always in the enabled-state.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs:182]: that's exactly it.  it also keeps things stateless so we don't need to keep track of everything as we walk.

File: src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.cs [hunk 11]
```diff
@@ -55,26 +53,7 @@ public async Task<Document> AddSourceToAsync(Document document, Compilation symb
             return await Simplifier.ReduceAsync(formattedDoc, reducers, null, cancellationToken).ConfigureAwait(false);
         }
 
-        /// <summary>
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.cs:58]: I looked into this and couldn't find any reason for it to exist :-/  it seems super strange.  Note: we only remove imports the compiler says are not used.  so, if we're removing imports we sholud be keeping, we should actually just fix the compiler there.

File: src/Workspaces/CSharp/Portable/CodeGeneration/AttributeGenerator.cs [hunk 12]
```diff
@@ -61,16 +74,39 @@ private static AttributeSyntax GenerateAttribute(AttributeData attribute, CodeGe
...
+        private static bool IsCompilerInternalAttribute(AttributeData attribute)
+        {
+            // from https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md
+            var attrClass = attribute.AttributeClass;
+            if (attrClass == null)
+                return false;
+
+            var name = attrClass.Name;
+
+            if (name != "NullableAttribute" && name != "NullableContextAttribute" && name != "NativeIntegerAttribute")
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/AttributeGenerator.cs:96]: agreed.  let me look into dynamic.

File: src/Workspaces/Core/Portable/CodeGeneration/NullableSyntaxAnnotation.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.CodeGeneration
+{
+    /// <summary>
+    /// Annotation placed if the code generator encounters a NullableAttribute or NullableContextAttribute while
+    /// generating the code.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeGeneration/NullableSyntaxAnnotation.cs:9]: sure

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs [hunk 14]
```diff
@@ -56,6 +58,22 @@ internal static partial class ITypeSymbolExtensions
                 syntax = syntax.WithAdditionalAnnotations(DoNotAllowVarAnnotation.Annotation);
             }
 
+#if !CODE_STYLE
+
+            if (type != null && type.IsReferenceType)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs:63]: will add tests.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs [hunk 15]
```diff
@@ -56,6 +58,22 @@ internal static partial class ITypeSymbolExtensions
...
+            {
+                syntax = syntax.WithAdditionalAnnotations(
+                    type.NullableAnnotation switch
+                    {
+                        NullableAnnotation.None => NullableSyntaxAnnotation.None,
+                        NullableAnnotation.Annotated => NullableSyntaxAnnotation.Annotated,
+                        NullableAnnotation.NotAnnotated => NullableSyntaxAnnotation.NotAnnotated,
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs:70]: it shouldn't  be necessary.  the reason being that you really can't have a mix of obliovus and non-oblivious.  you *can* have a mix of annotated/not-annotated.  but those are the same category here.  i.e. if i have `List<string?>` then the List is not-annotated, and the string is annotated.  But either of those will trigger `#nullable enable`.  

What triggers #nullable disable, is only i we then have *oblivious* values.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs:63]: tests added.

File: src/Workspaces/CSharp/Portable/CodeGeneration/AttributeGenerator.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/AttributeGenerator.cs]: fixed dynamic.

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcouv) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:594]: Tagging @cston. I'm surprised that we emit a `NullableAttribute(2)` (means `?`) on an unconstrained `T` in the first place ([sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFAACAzIQJZrwIBmEAxkQMIA8AKgD4qAbwCQVQlMLAA9rKyEAogEcMELFAAUAbQCCOWQHcActiwBdQv0KyYAC0QBKANxUAvlSYtSi6gCZCAyxjMxw9GBgEBmAMeEIQQmoARgA6CKiYuKIRQncxIA=))

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: done!

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2616]: done!

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs [hunk 16]
```diff
@@ -2352,9 +2336,744 @@ void M([|D|]&lt;int&gt; lambda)
...
+
+namespace N
+{
+    public class TestType
+    {
+        public void M1(string s)
+        {
+        }
+
+    #nullable disable
```
Comment (jcouv) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:2925]: nit: formatting

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs [hunk 17]
```diff
@@ -2352,9 +2336,744 @@ void M([|D|]&lt;int&gt; lambda)
...
+            var expected = $@"#region {FeaturesResources.Assembly} ReferencedAssembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
+// {CodeAnalysisResources.InMemoryAssembly}
+#endregion
+
+public class TestType
+{{
+    public TestType();
+
+    public void [|M1|](dynamic s);
+}}";
```
Comment (jcouv) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:3076]: Consider including a nasty case (where best effort fails):
```
List<string
#nullable enable
    > M(
#nullable disable
        string parameter) => throw null;
````

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cston) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs:594]: The attribute value represents the lack of a `notnull` constraint - see [examples](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFACWa8CAZhAMZECSAjADwAqAfFQDeVQqMIABACyEAYgAoehAPYwAFogCUAbioBfKrXpNWhNgCZefQgHd1CIopCE0KzDiEixU2QuVrNOpT6lEA===).

**Pull request merged**